test/gtest/ucp/test_ucp_sockaddr.cc

```diff
@@ -1117,62 +1132,72 @@ class test_ucp_sockaddr_protocols : public test_ucp_sockaddr {
 
-            if (recv_stop != NULL) {
+            if (recv_stop) {
                 sender().disconnect_nb(0, 0, UCP_EP_CLOSE_MODE_FORCE);
             } else {
-                send_stop(arg);
+                disconnect(*this, sender());
                 receiver().disconnect_nb(0, 0, UCP_EP_CLOSE_MODE_FORCE);
             }
         }
 
-        return ucp_tag_msg_recv_nb(receiver().worker(), &recv_buf[0],
-                                   size, ucp_dt_make_contig(1), message,
-                                   !err_handling ? rtag_complete_cb :
-                                   rtag_complete_err_handling_cb);
+        ucp_request_param_t recv_param = {};
+        recv_param.op_attr_mask        = UCP_OP_ATTR_FIELD_CALLBACK;
+        recv_param.cb.recv             = reinterpret_cast
+                                         <ucp_tag_recv_nbx_callback_t>(
+                                             !err_handling ? rtag_complete_cb :
+                                             rtag_complete_err_handling_cb);
+        return ucp_tag_msg_recv_nbx(receiver().worker(), &recv_buf[0], size, message,
+                                    &recv_param);
+    }
+
+    void sreq_release(void *sreq) {
+        if ((sreq != NULL) && UCS_PTR_IS_PTR(sreq)) {
+            if (ucp_request_check_status(sreq) == UCS_INPROGRESS) {
+                ucp_request_t *req = (ucp_request_t*)sreq - 1;
+                req->flags        |= UCP_REQUEST_FLAG_COMPLETED;
+
+                ucp_request_t *req_from_id =
+                    ucp_worker_get_request_by_id(sender().worker(),
+                                                 req->send.msg_proto.sreq_id);
+                if (req_from_id != NULL) {
+                    EXPECT_EQ(req, req_from_id);
+                    /* check PTR MAP flag only in this way, since it is debug
+                     * only flag has 0 value in a release mode */
+                    EXPECT_TRUE((req->flags & UCP_REQUEST_FLAG_IN_PTR_MAP) ==
+                                UCP_REQUEST_FLAG_IN_PTR_MAP);
+                    ucp_worker_del_request_id(sender().worker(), req,
+                                              req->send.msg_proto.sreq_id);
+                }
+            }
+
+            ucp_request_release(sreq);
+        }
     }
 
     void test_tag_send_recv(size_t size, bool is_exp,
                             bool is_sync = false, bool err_handling_test = false)
     {
-        stop_cb_t send_stop = (err_handling_test &&
-                               (get_variant_value() & SEND_STOP)) ?
-                              send_disconnect : NULL;
-        stop_cb_t recv_stop = (err_handling_test &&
-                               (get_variant_value() & RECV_STOP)) ?
-                              recv_disconnect : NULL;
+        bool send_stop     = err_handling_test && (get_variant_value() & SEND_STOP);
+        bool recv_stop     = err_handling_test && (get_variant_value() & RECV_STOP);
```

Comment:
when can we have SEND_STOP/RECV_STOP while err_handling_test == false?

Reply:
here:
https://github.com/openucx/ucx/blob/6a3dec007628ce383a822dea9991397071815d85/test/gtest/ucp/test_ucp_sockaddr.cc#L1631

Reply:
do we really expect killing the receiver will work ok without err handling support?

Reply:
> do we really expect killing the receiver/sender will work ok without err handling support?

we will not kill receiver/sender w/o error handling support.
SEND_STOP/RECV_STOP are variants that are set only by error-handling sockaddr test

Reply:
AFAIU, that contradicts what you wrote in https://github.com/openucx/ucx/pull/6001#discussion_r555005549

Reply:
> AFAIU, that contradicts what you wrote in [#6001 (comment)](https://github.com/openucx/ucx/pull/6001#discussion_r555005549)

sorry, maybe it was not clear
https://github.com/openucx/ucx/blob/6a3dec007628ce383a822dea9991397071815d85/test/gtest/ucp/test_ucp_sockaddr.cc#L1631
is method of `test_ucp_sockaddr_protocols_err` which extends `test_ucp_sockaddr_protocols` with error handling.
but before testing error handling, we have to do warmup w/o killing sender/receiver and then real error handling test

Reply:
separated the flags to `send_stop`/`recv_stop` to make it clear as suggested by you